File: packages/typescript/src/types/service.ts
import type { Metadata } from "./primitives";
import type { JsonValue } from "./proto.js";
import type { IAgentRuntime } from "./runtime";

/**
 * Core service type registry that can be extended by plugins via module augmentation.
 * Plugins can extend this interface to add their own service types:
 *
 * @example
 * ```typescript
 * declare module '@elizaos/core' {
 *   interface ServiceTypeRegistry {
 *     MY_CUSTOM_SERVICE: 'my_custom_service';
 *   }
 * }
 * ```
 */
export interface ServiceTypeRegistry {
	TRANSCRIPTION: "transcription";
	VIDEO: "video";
	BROWSER: "browser";
	PDF: "pdf";
	REMOTE_FILES: "aws_s3";
	WEB_SEARCH: "web_search";
	EMAIL: "email";
	TEE: "tee";
	TASK: "task";
	APPROVAL: "approval";
	TOOL_POLICY: "tool_policy";
	WALLET: "wallet";
	LP_POOL: "lp_pool";
	TOKEN_DATA: "token_data";
	MESSAGE_SERVICE: "message_service";
	MESSAGE: "message";
	POST: "post";
	HOOKS: "hooks";
	PAIRING: "pairing";
	AGENT_EVENT: "agent_event";
	OPTIMIZED_PROMPT: "optimized_prompt";
	UNKNOWN: "unknown";
}

/**
 * Type for service names that includes both core services and any plugin-registered services
 */
export type ServiceTypeName = ServiceTypeRegistry[keyof ServiceTypeRegistry];

/**
 * Helper type to extract service type values from the registry
 */
export type ServiceTypeValue<K extends keyof ServiceTypeRegistry> =
	ServiceTypeRegistry[K];

/**
 * Helper type to check if a service type exists in the registry
 */
export type IsValidServiceType<T extends string> = T extends ServiceTypeName
	? true
	: false;

/**
 * Type-safe service class definition
 */
export type TypedServiceClass<T extends ServiceTypeName> = {
	new (runtime?: IAgentRuntime): Service;
	serviceType: T;
	start(runtime: IAgentRuntime): Promise<Service>;
};

/**
 * Map of service type names to their implementation classes.
 * Plugins can extend this via module augmentation:
 * @example
 * ```typescript
 * declare module '@elizaos/core' {
 *   interface ServiceClassMap {
 *     MY_SERVICE: typeof MyService;
 *   }
 * }
 * ```
 */
// biome-ignore lint/complexity/noBannedTypes: Empty interface for module augmentation
export type ServiceClassMap = {};

/**
 * Helper to infer service instance type from service type name
 */
export type ServiceInstance<T extends ServiceTypeName> =
	T extends keyof ServiceClassMap ? InstanceType<ServiceClassMap[T]> : Service;

/**
 * Runtime service registry type
 */
export type ServiceRegistry<T extends ServiceTypeName = ServiceTypeName> = Map<
	T,
	Service
>;

/**
 * Enumerates the recognized types of services that can be registered and used by the agent runtime.
 * Services provide specialized functionalities like audio transcription, video processing,
 * web browsing, PDF handling, file storage (e.g., AWS S3), web search, email integration,
 * secure execution via TEE (Trusted Execution Environment), and task management.
 * This constant is used in `AgentRuntime` for service registration and retrieval (e.g., `getService`).
 * Each service typically implements the `Service` abstract class or a more specific interface like `IVideoService`.
 */
export const ServiceType = {
	TRANSCRIPTION: "transcription",
	VIDEO: "video",
	BROWSER: "browser",
	PDF: "pdf",
	REMOTE_FILES: "aws_s3",
	WEB_SEARCH: "web_search",
	EMAIL: "email",
	TEE: "tee",
	TASK: "task",
	APPROVAL: "approval",
	TOOL_POLICY: "tool_policy",
	WALLET: "wallet",
	LP_POOL: "lp_pool",
	TOKEN_DATA: "token_data",
	MESSAGE_SERVICE: "message_service",
	MESSAGE: "message",
	POST: "post",
	HOOKS: "hooks",
	PAIRING: "pairing",
	AGENT_EVENT: "agent_event",
	VOICE_CACHE: "voice_cache",
	OPTIMIZED_PROMPT: "optimized_prompt",
	UNKNOWN: "unknown",
} as const;

/**
 * Client instance
 */
export abstract class Service {
	/** Runtime instance */
	protected runtime!: IAgentRuntime;

	constructor(runtime?: IAgentRuntime) {
		if (runtime) {
			this.runtime = runtime;
		}
	}

	abstract stop(): Promise<void>;

	/** Service type */
	static serviceType: string;

	/** Service name */
	abstract capabilityDescription: string;

	/** Service configuration */
	config?: Metadata;

	/** Start service connection - subclasses must override this */
	static async start(_runtime: IAgentRuntime): Promise<Service> {
		throw new Error("Service.start() must be implemented by subclass");
	}

	/** Stop service connection - optional, subclasses may override this */
	static stopRuntime?(_runtime: IAgentRuntime): Promise<void>;

	/** Optional static method to register send handlers */
	static registerSendHandlers?(runtime: IAgentRuntime, service: Service): void;
}

/**
 * Generic service interface that provides better type checking for services
 * @template ConfigType The configuration type for this service
 * @template InputType The input type for processing
 * @template ResultType The result type returned by the service operations
 */
export interface TypedService<
	ConfigType extends Metadata = Metadata,
	InputType = JsonValue,
	ResultType = JsonValue,
> extends Service {
	/**
	 * The configuration for this service instance
	 */
	config?: ConfigType;

	/**
	 * Process an input with this service
	 * @param input The input to process
	 * @returns A promise resolving to the result
	 */
	process(input: InputType): Promise<ResultType>;
}

/**
 * Generic factory function to create a typed service instance.
 * getService() is synchronous — no await needed.
 * @param runtime The agent runtime
 * @param serviceType The type of service to get
 * @returns The service instance or null if not available
 */
export function getTypedService<
	ConfigType extends Metadata = Metadata,
	InputType = JsonValue,
	ResultType = JsonValue,
>(
	runtime: IAgentRuntime,
	serviceType: ServiceTypeName,
): TypedService<ConfigType, InputType, ResultType> | null {
	return runtime.getService<TypedService<ConfigType, InputType, ResultType>>(
		serviceType,
	);
}

/**
 * Standardized service error type for consistent error handling
 */
export interface ServiceError {
	code: string;
	message: string;
	details?: Record<string, JsonValue> | string | number | boolean | null;
	cause?: Error;
}

/**
 * Safely create a ServiceError from any caught error
 */
export function createServiceError(
	error: Error | string | JsonValue,
	code = "UNKNOWN_ERROR",
): ServiceError {
	if (error instanceof Error) {
		return {
			code,
			message: error.message,
			cause: error,
		};
	}

	return {
		code,
		message: String(error),
	};
}

File: packages/rust/src/types/service.rs
//! Service types for elizaOS
//!
//! Contains Service trait, service types, and related interfaces.

use serde::{Deserialize, Serialize};

use crate::platform::PlatformService;

use super::primitives::Metadata;

/// Core service type names
pub mod service_type {
    /// Transcription service type
    pub const TRANSCRIPTION: &str = "transcription";
    /// Video service type
    pub const VIDEO: &str = "video";
    /// Browser service type
    pub const BROWSER: &str = "browser";
    /// PDF service type
    pub const PDF: &str = "pdf";
    /// Remote files service type (AWS S3)
    pub const REMOTE_FILES: &str = "aws_s3";
    /// Web search service type
    pub const WEB_SEARCH: &str = "web_search";
    /// Email service type
    pub const EMAIL: &str = "email";
    /// TEE (Trusted Execution Environment) service type
    pub const TEE: &str = "tee";
    /// Task service type
    pub const TASK: &str = "task";
    /// Wallet service type
    pub const WALLET: &str = "wallet";
    /// LP pool service type
    pub const LP_POOL: &str = "lp_pool";
    /// Token data service type
    pub const TOKEN_DATA: &str = "token_data";
    /// Message service type
    pub const MESSAGE_SERVICE: &str = "message_service";
    /// Message service type (alias)
    pub const MESSAGE: &str = "message";
    /// Post service type
    pub const POST: &str = "post";
    /// Hooks service type
    pub const HOOKS: &str = "hooks";
    /// Unknown service type
    pub const UNKNOWN: &str = "unknown";
}

/// Service definition for serialization
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceDefinition {
    /// Service type name
    pub service_type: String,
    /// Capability description
    pub capability_description: String,
    /// Optional configuration
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config: Option<Metadata>,
}

/// Service trait for all services
#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
pub trait Service: PlatformService {
    /// Get the service type
    fn service_type(&self) -> &str;

    /// Get the capability description
    fn capability_description(&self) -> &str;

    /// Get the service configuration
    fn config(&self) -> Option<&Metadata> {
        None
    }

    /// Stop the service
    async fn stop(&self) -> Result<(), anyhow::Error>;
}

/// Typed service trait for services with specific input/output types
#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
pub trait TypedService<Input, Output>: Service {
    /// Process an input
    async fn process(&self, input: Input) -> Result<Output, anyhow::Error>;
}

/// Service error type
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceError {
    /// Error code
    pub code: String,
    /// Error message
    pub message: String,
    /// Additional details
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
}

impl ServiceError {
    /// Create a new service error
    pub fn new(code: &str, message: &str) -> Self {
        ServiceError {
            code: code.to_string(),
            message: message.to_string(),
            details: None,
        }
    }

    /// Create from an error
    pub fn from_error(error: impl std::error::Error, code: &str) -> Self {
        ServiceError {
            code: code.to_string(),
            message: error.to_string(),
            details: None,
        }
    }
}

impl std::fmt::Display for ServiceError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[{}] {}", self.code, self.message)
    }
}

impl std::error::Error for ServiceError {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_service_definition_serialization() {
        let def = ServiceDefinition {
            service_type: "transcription".to_string(),
            capability_description: "Audio transcription service".to_string(),
            config: None,
        };

        let json = serde_json::to_string(&def).unwrap();
        assert!(json.contains("\"serviceType\":\"transcription\""));
        assert!(json.contains("\"capabilityDescription\""));
    }

    #[test]
    fn test_service_error_display() {
        let error = ServiceError::new("NOT_FOUND", "Resource not found");
        assert_eq!(error.to_string(), "[NOT_FOUND] Resource not found");
    }
}
